docs/development/design/build-images.rst

```diff
@@ -0,0 +1,284 @@
+
+In case we need to add an extra requirement to the ``base`` image,
+we will need to rebuild all of them.
+The new image *may have different package versions* since there may be updates on the Ubuntu repositories.
+This conveys some small risk here, but in general we shouldn't require to add packages to the base images.
+
+Users with specific requirements could use ``build.system_packages`` and/or ``build.extras`` in the config file.
+
+How do we remove an old Python version?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+At some point an old version of Python will be deprecated (eg. 3.4) and will be removed.
+To achieve this, we can just remove the Docker image affected: ``ubuntu20-py34``,
+once there are no users depending on it anymore.
+
+We will know which projects are using these images because they are pinning it in the config file.
+We could show a message in the build output page and also send them an email with the EOL date for this image.
+
+Deprecation plan
+----------------
+
+It seems we have ~50Gb free on builders disks.
+Considering that the new images will be sized approximately (built locally as test):
+
+* ``ubuntu20-base``: ~5Gb
+* ``ubuntu20-py27``: ~150Mb
+* ``ubuntu20-py36``: ~210Mb
+* ``ubuntu20-py39``: ~20Mb
+* ``ubuntu20-conda47``: ~713Mb
+
+which is about ~6Gb in total, we still have plenty of space.
+
+We could keep ``stable``, ``latest`` and ``testing`` for some time without worry too much.
+New projects shouldn't be able to select these images and they will be forced to use ``ubuntu20``
+if they don't specify one.
+
+We may want to keep the two latest Ubuntu LTS releases available in production.
+At the moment of writing this they are:
+
+* Ubuntu 18.04 LTS (our ``stable``, ``latest`` and ``testing`` images)
+* Ubuntu 20.04 LTS (our new ``ubuntu20``)
+
+Once Ubuntu 22.04 LTS is released, we should deprecate Ubuntu 18.04 LTS,
+and give users 6 months to migrate to a newer image.
+
+
+Work required
+-------------
+
+There are a lot of work to do here.
+However, we want to prioritize it based on users' impact.
+
+#. allow users to install packages with APT
+
+   * update config file to support ``build.system_packages`` config
+   * modify builder code to run ``apt-get install`` as ``root`` user
+
+#. allow users to install extras via config file
+
+   * update config file to support ``build.extras`` config
+   * modify builder code to run ``nodenv install`` / ``rustup install``
+   * re-build our current images with pre-installed nodenv and rustup
+   * make sure that all the versions are the same we have in production
+   * deploy builders with newer images
+
+#. pre-build commands (not covered in this document)
+
+#. new structure
+
+   * update config file to support new image names for ``build.image``
+   * automate Docker image building
+   * deploy builders with newer images
+
+
+Conclusion
+----------
+
+I don't think we need to differentiate the images by its state (stable, latest, testing)
```

Comment:
Tiny comment, perhaps because this is the first design document I read in depth - it was weird finding a sentence in first person here 😅 Was it intentional?

Reply:
Not really. I can change it for "There is no need to..."

Reply:
It seems it's the only place where I used first person in the document :upside_down_face: